scarb/src/ops/resolve.rs

```diff
@@ -123,6 +96,28 @@ pub fn generate_compilation_units(
             }
         });
 
+        let current_version = crate::version::get().cairo.version.to_version().unwrap();
+        let matching_version = packages.iter().all(|pkg| {
+            let package_version = &pkg.manifest.metadata.cairo_version;
+            let mut is_matching = true;
+            if let Some(package_version) = package_version  {
+                is_matching = package_version.matches(&current_version);
+                if !is_matching {
+                    ws.config().ui().error(format!(
+                        "Package {}. Required Cairo version isn't compatible with current version. Should be: {} is: {}",
+                        pkg.id.name, package_version.to_string(), current_version.to_string()
+                    ));
+                }
+            }
+            is_matching
```

Comment:
protip, this would be more idiomatic I think:

```suggestion
            match &pkg.manifest.metadata.cairo_version {
                Some(package_version) if !package_version.matches(&current_version) => {
                    ws.config().ui().error(format!(
                        "Package {}. Required Cairo version isn't compatible with current version. Should be: {} is: {}",
                        pkg.id.name, package_version, current_version
                    ));
                    false
                }
                _ => true
            }
```

note that I have removed `to_string()` calls, because they come from `ToString` trait which is implemented if a type implements `Display`, and `format!` macro can happily consume `Display` types